### DISCUSSION TITLE
Should "useFieldArray append" types behave the same as the defaultValues type?

### DISCUSSION THREAD

**Discussion by mauriciosoares** [react-hook-form/react-hook-form#10211] [Q&A]
When using an array of objects with react-hook-form, the default values accepts that the keys from the object can be undefined, even though that the schema says that these are mandatory. And that's great since I can initialize my form with an empty state:

![image](https://user-images.githubusercontent.com/2321259/229148606-f1c3b380-a0e8-46cc-b304-59f7fbfe8a92.png)

However, when using `append` from the `useFieldArray` method, I'm forced to set a value for the keys of the object I'm appending into the form. For me that's a problem because I want to append a value that's completely empty, the same way as the default value, so the user can add the initial data on it

![image](https://user-images.githubusercontent.com/2321259/229148965-53d42cf0-ea4c-4bb5-a5ee-11bdfc50ae98.png)

So my question is, should the append behave the same as the default values? Allowing us to add an empty value for the initial appended data?

Here's a code sandbox with an example:

https://codesandbox.io/s/youthful-faraday-feem73?file=/src/index.tsx

Thanks!

**Comment by leapful** [comment]
@mauriciosoares 

react-hook-form allows to input partial default values because there is always a reasonable case when providing full default values object is not needed, therefore on all fields passed to `defaultValues` can be optional.

`append` method is not the same because you should know exactly what should be appended at the time triggering `append`.

If you want to make it consistent, I would recommend to use empty string as default value for `useForm` hook and also for `append` method.

**Reply by mauriciosoares** [reply]
Well, I came across another scenario that's not specific to strings, but is also related to the types on `append` vs `defaultValues`.

Consider that we have a field that is a non empty array, lets say `tickets`. we'd type it as a `nonempty` array with zod so react-hook-form can throw errors if the user tries to submit the form if that array is empty.

When we set the default values, we can set it as empty, because this might be a user input based field, they'll add an array of ticket numbers that only they are aware of...

However when appending a new item to the array list, now I can't set it as an empty array, because of the `nonempty` type. And I can't really fill it in with a random empty string, because that'd break the form validation. I updated the codesandbox example with the code for that scenario above

https://codesandbox.io/s/youthful-faraday-feem73?file=/src/index.tsx

> append method is not the same because you should know exactly what should be appended at the time triggering append.

I think that is not a reality in practice, the append method is used to add a brand new field to the form for the user to type his info, having to prefilling it with data, even if it's just dummy data as a string in an array, is not always possible.

**Reply by andho** [reply]
> `append` method is not the same because you should know exactly what should be appended at the time triggering `append`

It's not possible to know exactly what should be appended because we are expecting the user to provide this info. When the fields are displayed after appending, the user will input the values and at that point we will know the exact values. But to come to this point, we already need to call `append`. Is this not the use case for append?

**Comment by thiago-silva-ac** [comment]
Having same issue but mine is with dates, how can i create a empty Date object? With default value I used undefined but append doesn't accept it.

**Reply by andho** [reply]
Same thing with numbers.

**Comment by juno7803** [comment]
Having same issue here. I want to append form by same type of defaultValues. (empty state)

**Comment by hueindahaus** [comment]
I don't understand the design choice of having to know exactly what should be filled in when adding another form component...

**Comment by 25747** [comment]
i'm having the same issues with dates and numbers. i am trying to work around it by tweaking some type definitions and Yup schema, but it isn't pretty and breaks our existing form patterns.

**Comment by wadeamaral** [comment]
Same issue here. Allowing for undefined values in append would be extremely helpful.

**Comment by lazcanoluca** [comment]
Same, it doesn't seem right that `append` requires valid default values. I believe it'd be much easier and consistent would they behave like `defaultValues`, unless I'm missing something.

**Comment by coded58** [comment]
seems this hasn't been resolved? same issue with @thiago-silva-ac, appending both date and numbers. Having a number input and when user appends we default it to 0 doesn't look great, same with dates

**Comment by mauriciosoares** [comment]
Hey @bluebill1049, as you're the official maintainer of react-hook-form I wanted to tag you in this discussion so you can give your thoughts. Quite a few people are having the same issue that I brought up initially, is there any major drawbacks on making the `append` method accept `Partial` values from the schema like the defaultValues?

**Reply by perevezencev** [reply]
+1

**Comment by barbenezra** [comment]
+1 on this one
would be very helpful

**Comment by ImHamba** [comment]
+1, it makes no sense to append fields to a form that are meant to be filled by the user (e.g. a user defined number of entries to fill) but expect to have the data already to include in the append call

**Comment by aq1018** [comment]
This is really a questionable design choice. This is how I worked around it in typescript though. It's ugly, but it works

```ts
append({
  name: '', // string, give it empty string
  type: '' as SomeEnum, // enum, use empty string and type cast to enum
  amount: '' as unknown as number, // number, lie to TS basically,
  date: '' as unknown as Date, // date object, just keep lying to TS... :(
})
```

**Reply by salmanbabri** [reply]
I was recently working on an old project with controlled component forms. This is exactly what I did in the initial state.

Its kinda ugly at first. But I made it a bit less distracting by extracting empty states into a utils file like this:-

`export const emptyNumber = "" as unknown as number;`

**Reply by SimonVillage** [reply]
This worked for me. It took me a few hours to finally find a solution for this.

**Reply by braaar** [reply]
I did the same thing. Sometimes type casting is the least painful solution.

**Reply by MatthieuJnon** [reply]
This is not type casting, type casting changes the type of the variable, this is a type assertion, the type stays wrong.

**Comment by iffa** [comment]
Here's a patch compatible with version `7.52.1` for those that want to take this approach instead:

```diff
diff --git a/node_modules/react-hook-form/dist/types/fieldArray.d.ts b/node_modules/react-hook-form/dist/types/fieldArray.d.ts
index 5fa9572..db63797 100644
--- a/node_modules/react-hook-form/dist/types/fieldArray.d.ts
+++ b/node_modules/react-hook-form/dist/types/fieldArray.d.ts
@@ -2,6 +2,7 @@ import { FieldValues } from './fields';
 import { Control } from './form';
 import { FieldArrayPath, FieldArrayPathValue } from './path';
 import { RegisterOptions, Validate } from './validator';
+import { DeepPartial } from './utils';
 export type UseFieldArrayProps<TFieldValues extends FieldValues = FieldValues, TFieldArrayName extends FieldArrayPath<TFieldValues> = FieldArrayPath<TFieldValues>, TKeyName extends string = 'id'> = {
     name: TFieldArrayName;
     keyName?: TKeyName;
@@ -97,7 +98,7 @@ export type UseFieldArrayPrepend<TFieldValues extends FieldValues, TFieldArrayNa
  * </button>
  * ```
  */
-export type UseFieldArrayAppend<TFieldValues extends FieldValues, TFieldArrayName extends FieldArrayPath<TFieldValues> = FieldArrayPath<TFieldValues>> = (value: FieldArray<TFieldValues, TFieldArrayName> | FieldArray<TFieldValues, TFieldArrayName>[], options?: FieldArrayMethodProps) => void;
+export type UseFieldArrayAppend<TFieldValues extends FieldValues, TFieldArrayName extends FieldArrayPath<TFieldValues> = FieldArrayPath<TFieldValues>> = (value: FieldArray<DeepPartial<TFieldValues>, TFieldArrayName> | FieldArray<TFieldValues, TFieldArrayName>[], options?: FieldArrayMethodProps) => void;
 /**
  * Remove field/fields at particular position.
  *
```

**Reply by nuclei272** [reply]
This fixed it for me. Thanks!

**Reply by TheDevick** [reply]
Apparently this is the only way to solve the issue, thanks!

But modifying `node_modules` folder is, of course, far from ideal.

@iffa could you open a PR to the repository to fix that?

**Comment by gergelykisshun** [comment]
I had the same problem. 

I resorted to adding .nullable() in zod to the field and use null in append() for the field. 

This caused the field not to be required anymore, so I added a superRefine to check for the field and make it required. Super hacky... would be solved if append would work the same as defaultValues and can accept Partial values.

**Comment by Gustavo-Pauli** [comment]
Did someone created an issue for this?

**Comment by TheDevick** [comment]
Yep, this issue seems to persist.

I’m currently trying to display placeholder text in a select input, but it’s basically impossible because I need the value to be defined, and apparently, the only way to display the placeholder is setting the value as null.

So, @bluebill1049 , can you help us here? Thanks!

**Comment by daniel-barhanin** [comment]
I'm having the same problem, will there soon be a fix for it? @iffa will this patch be deployed?

**Comment by SimonVillage** [comment]
Having the same problem. I don't really understand why it is expected to provide defined values to append if we don't know what the user intends to fill into the form? Especially with numbers or enum type of fields. Providing a 0 as default isn't an option, same for enum type of select fields, providing a default select value doesn't make sense.

**Comment by bismitpanda** [comment]
It has been an issue for nearly 2 years. Why is there still no **official** (may not be the right word, but get the idea) solution to this?

**Comment by valtron** [comment]
I'm using:
```typescript
import * as rhf from 'react-hook-form';
import type {
	DeepPartial, UseFieldArrayProps, UseFieldArrayReturn,
	FieldValues, FieldArrayPath, FieldArrayMethodProps, FieldArray,
} from 'react-hook-form';

export function usePartialFieldArray<
	TFieldValues extends FieldValues,
	TFieldArrayName extends FieldArrayPath<TFieldValues> = FieldArrayPath<TFieldValues>,
>(props: UseFieldArrayProps<TFieldValues, TFieldArrayName>)
: Omit<UseFieldArrayReturn<TFieldValues, TFieldArrayName>, 'append'> & {
	append: (value: DeepPartial<FieldArray<TFieldValues, TFieldArrayName>>, options?: FieldArrayMethodProps) => void;
} {
	const { append, ...rest } = rhf.useFieldArray(props);
	function appendPartial(value: DeepPartial<FieldArray<TFieldValues, TFieldArrayName>>, options?: FieldArrayMethodProps): void {
		append(value as FieldArray<TFieldValues, TFieldArrayName>, options);
	}
	return { ...rest, append: appendPartial };
}
```

A similar problem exists with `setValue`, btw.

**Reply by andho** [reply]
Yes, same issue with `setValue`, but you can at least use `resetField` instead
